### ISSUE TITLE
Question about the implement of Sphere embedding

### ISSUE THREAD

**Issue by shawjan** [davidsandberg/facenet#119]
Hi! David,
    thanks for your code so that I could understand the paper FaceNet further. But I have a question about sphere embedding in the paper. As introduced in this paper, the embedding layer just like a fully connected layer, which you can get a low dimensional feature by formulation Y = WX + b(in this formulation, X means input feature, Y means output feature, W means weight, b means bias). In my opinion, the embedding layer seems to have another constraint like ||Y||_{2}=1, which means to constrain embedding to live on the d-dimension hypersphere. However, I did't find this constraint in your code. And I know your code referred to OpenFace, which I also did't find.  Meanwhile, in your implement, L2 operates after Embedding, which is different from structure in the paper. 
    As your code said, is that means the output of CNN can directly calculate Euclidean distance for similarity? Why this method haven't been told until FaceNet released? And why FaceNet want to emphasis on the words below:

> The embedding is represented by f (x) ∈ Rd . It embeds an image x into a d-dimensional Euclidean space. Additionally, we constrain this embedding to live **on the d-dimensional hypersphere**, i.e. ∥f(x)∥2 = 1. 

Those questions confuse me a lot, Looking forward to your answer. Thanks a lot!

**Comment by davidsandberg** [comment]
Hi @shawjan!
Short answer to a long question: The L2 normalization is forcing the embedding to live on a hypersphere, e.g. the line
```
embeddings = tf.nn.l2_normalize(pre_embeddings, 1, 1e-10, name='embeddings')
```
in `facenet_train.py`
I don't think there is anything like an embedding layer as such, but rather the activations from a previous layer are normalized and used as an embedding, but that's a matter of how you see it I guess. As you mention, the nice thing with the embeddings is that distances directly correspond to a measure of face similarity.

**Comment by shawjan** [comment]
Thanks for your answer！Maybe I misunderstand L2 normalization so that I think there should be a embedding layer.

**Comment by remarknibor** [comment]
Hi David,
I was wondering if there was a way to convert the L2 normalized embedding for a given face back to the raw embedding vector representing the location in the Euclidean space?
Thanks,
Robin.
